### ISSUE TITLE
Remove dependency on OpenCV for capture and encoding

### ISSUE THREAD

**Issue by dzhoshkun** [gift-surg/GIFT-Grab#9]
GIFT-Grab uses OpenCV for:
* [Xvid encoding / ~~decoding~~](https://github.com/gift-surg/GIFT-Grab/blob/master/doc/build.md#xvid)
* Capturing BGRA frames from [Epiphan DVI2PCIe Duo](https://github.com/gift-surg/GIFT-Grab/blob/master/doc/build.md#epiphan-dvi2pcie-duo) and [network streams](https://github.com/gift-surg/GIFT-Grab/blob/master/doc/build.md#network-streams)

However OpenCV depends on FFmpeg, presumably for these tasks. FFmpeg is already used in GIFT-Grab as well. And OpenCV is quite a heavy-weight dependency in this regard. So it might be worth removing it in favour of using FFmpeg for all that OpenCV is needed for in GIFT-Grab.

With this, the OpenCV dependency will become optional, and for the time being confined to #10

**Comment by tvercaut** [comment]
That makes sense. One thing we discussed a long time ago though was to be able to decorate/adapt a VideoFrame (https://github.com/gift-surg/GIFT-Grab/blob/master/src/api/videoframe.h) to present it as an OpenCV image if OpenCV is specified as a dependency. This would ease integrating GIFT-Grab in processing pipelines.

**Comment by dzhoshkun** [comment]
* [x] depends on #7

**Comment by dzhoshkun** [comment]
A [more thorough search](https://github.com/gift-surg/GIFT-Grab/search?p=1&q=opencv&unscoped_q=opencv), which:

* [ ] should be revisited (for repo items to be updated) before closing this issue,

 has revealed that OpenCV is also used for [capturing BGRA frames from Epiphan frame grabbers](https://github.com/gift-surg/GIFT-Grab/blob/87a0e922cb8174b6648fbfdb82849d4ffc9569b9/src/api/videosourcefactory.cpp#L67). 

* [ ] so BGRA captures from Epiphan should be supported by an appropriate alternative.

For instance Epiphan SDK provides the `V2U_GRABFRAME_FORMAT_ARGB32` flag (i.e. BGRA in reverse order). I reckon libVLC will provide an equivalent as well. However BGRA does not seem to be available. Maybe:

* [ ] check how this is implemented in OpenCV (as OpenCV uses video4linux under the hood, AFAIK).

**Comment by dzhoshkun** [comment]
@tvercaut I'm going to address OpenCV compatibility in a separate issue #10

**Comment by dzhoshkun** [comment]
[This thread](https://www.raspberrypi.org/forums/viewtopic.php?f=43&t=79422) might be useful for supporting BGRA capture with libVLC.

**Comment by dzhoshkun** [comment]
A video-4-linux example https://linuxtv.org/downloads/v4l-dvb-apis/uapi/v4l/v4l2grab.c.html potentially useful for https://github.com/gift-surg/GIFT-Grab/issues/9#issuecomment-405555893

**Comment by dzhoshkun** [comment]
There's [OpenCV color conversion involved in the current implementation](https://github.com/gift-surg/GIFT-Grab/blob/4dc37b7a4b4bec47a9c5747bcc0eefea2d0a1f7c/src/opencv/opencv_video_source.cpp#L125), so we could as well use the Epiphan SDK's `V2U_GRABFRAME_FORMAT_ARGB32` (https://github.com/gift-surg/GIFT-Grab/issues/9#issuecomment-405555893) and do the conversion manually!

**Comment by dzhoshkun** [comment]
[OpenCV seems to use loops for color conversion](https://github.com/opencv/opencv/blob/b39cd06249213220e802bb64260727711d9fc98c/modules/imgproc/src/color_rgb.cpp#L17)

**Comment by dzhoshkun** [comment]
* [ ] run CI up to above commit with an Epiphan device!
* [ ] more restructuring is needed e.g. in CMake configuration

**Comment by dzhoshkun** [comment]
FFmpeg seems to nail it:

```
[0] % ./tests/rgbswap/argb_to_bgra_profiler     
Profiling ARGB => BGRA conversion for 1920 x 1080 image
Loop (success) took: 11333 usec
memcpy took: 3739 usec
Function (success) took: 35876 usec
FFmpeg (success) took: 10 usec
```

**Comment by tvercaut** [comment]
As per #10, OpenCV compatibility is an important feature for many (but not all) users. Should we rename this issue as "Make OpenCV dependency optional" or "Remove mandatory core dependency on OpenCV"?

**Comment by dzhoshkun** [comment]
@tvercaut Agreed, I've updated the title and the description to make this issue more specific.
